add avatar caching for unknown users

lun-4 opened the pull request:
(pull request message taken from its single commit, which should
explain why we need this pr. if there are implementation questions,
feel free to ask)

it is possible for users to not share any guilds with
the bot. so we need to resort to something like get_user_info
but it has a low ratelimit. caching the avatar urls given
by get_user_info is a good strategy to circumventing it.

note: if the user already shares guilds with the bot, it won't
hit the cache at all.

the obvious downside is that avatar changes will take a while to
appear in the bot (considering the user still doesn't share any
guilds with it).

cache invalidation is set to 30 minutes, which should be a reasonable
default.

the cache is just a single dictionary because redis is way too much
for this kind of purpose.

lun-4 committed:
add avatar caching for unknown users

it is possible for users to not share any guilds with
the bot. so we need to resort to something like get_user_info
but it has a low ratelimit. caching the avatar urls given
by get_user_info is a good strategy to circumventing it.

note: if the user already shares guilds with the bot, it won't
hit the cache at all.

the obvious downside is that avatar changes will take a while to
appear in the bot (considering the user still doesn't share any
guilds with it).

cache invalidation is set to 30 minutes, which should be a reasonable
default.

the cache is just a single dictionary because redis is way too much
for this kind of purpose.

aveao commented:
`CACHE_INVALID = 30 * 60`

wouldn't it be better to pull this from the config yaml?

lun-4 commented:
yeah. will add it in

lun-4 committed:
add discord.avatar_cache configuration option

suggested by @aveao.

instance managers should be able to tweak this value depending on their
needs. the default of 30 minutes is also there

lun-4 committed:
use dict.get for configurations without avatar_cache
